CSharp/Library/Microsoft.Bot.Connector.Shared/Authentication/AdalAuthenticator.cs

```diff
@@ -0,0 +1,115 @@
+﻿using Microsoft.Extensions.Logging;
+using Microsoft.IdentityModel.Clients.ActiveDirectory;
+using System;
+using System.Collections.Generic;
+using System.Net.Http.Headers;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Bot.Connector.Shared.Authentication
+{
+    public class AdalAuthenticator 
+    {
+        private static readonly TimeSpan SemaphoreTimeout = TimeSpan.FromSeconds(1);
+        private const int MaxRetries = 5;
+
+        private static AuthenticationContext authContext = new AuthenticationContext(JwtConfig.ToChannelFromBotLoginUrl) { ExtendedLifeTimeEnabled = true };
+        private static SemaphoreSlim authContextSemaphore = new SemaphoreSlim(50, 50);
```

Comment:
A comment here would be good. I saw this and concluded the authContext needed to be protected in a standard thread-safe manner (Monitor / Mutex). Instead, this is being used as a throttle...